8051591: Test javax/swing/JTabbedPane/8007563/Test8007563.java fails

kumarabhi006 opened the pull request:
Test was failing on multiple platforms mainly windows and linux and was problemlisted for all platforms. Increased the delay from 200ms to 500 ms to fetch the correct background color using robot. Now test passed for multiple iteration (~50 times) in CI pipeline for all platforms. Wild imports are expanded and test is pulled out of an extra level of folder.
Test removed from `Problem-list` file as well.

CI testing link is posted in JBS.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8051591](https://bugs.openjdk.org/browse/JDK-8051591): Test javax/swing/JTabbedPane/8007563/Test8007563.java fails (**Bug** - P3)


### Reviewers
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - **Reviewer**)
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) Review applies to [1ff306c4](https://git.openjdk.org/jdk/pull/20791/files/1ff306c44788fcfa7f7a2392c1bb5921b34e1a74)
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**) Review applies to [1ff306c4](https://git.openjdk.org/jdk/pull/20791/files/1ff306c44788fcfa7f7a2392c1bb5921b34e1a74)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20791/head:pull/20791` \
`$ git checkout pull/20791`

Update a local copy of the PR: \
`$ git checkout pull/20791` \
`$ git pull https://git.openjdk.org/jdk.git pull/20791/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20791`

View PR using the GUI difftool: \
`$ git pr show -t 20791`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20791.diff">https://git.openjdk.org/jdk/pull/20791.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20791#issuecomment-2320812812)

honkar-jdk reviewed (commented):
```
       SecondaryLoop secondaryLoop =
                Toolkit.getDefaultToolkit().getSystemEventQueue()
                        .createSecondaryLoop();
```
I'm not sure if the SecondaryLoop is requiredfor this test or if it can be simplified?

kumarabhi006 commented:
> I'm not sure if the SecondaryLoop code is required for this test or it can be simplified?

I have tried simplifying the test code and removed SecondaryLoop code. Tested the current changes in CI across all platforms and no failure observed.

kumarabhi006 commented:
@prsadhuk As we were discussing about the test change and removal of SecondaryLoop change, I tested the current changes without the fix proposed in [JDK-8007563](https://bugs.openjdk.org/browse/JDK-8007563) and test failed which is correct. After applying the fix, test is able to pass.

kumarabhi006 commented:
@prsadhuk Ran the test with uiscale of 2 and test passed across all platforms.

prsadhuk commented:
> @prsadhuk As we were discussing about the test change and removal of SecondaryLoop change, I tested the current changes without the fix proposed in [JDK-8007563](https://bugs.openjdk.org/browse/JDK-8007563) and test failed which is correct. After applying the fix, test is able to pass.

Secondary loop was added for JDK-8078269 not for JDK-8007563 so you need to check if test fail without Secondary loop without that fix and passes with the fix..

kumarabhi006 commented:
> Secondary loop was added for JDK-8078269 not for JDK-8007563 so you need to check if test fail without Secondary loop without that fix and passes with the fix..

I tested without the fix for JDK-8078269 for different scenarion:

1. Old test i.e. with SecondaryLoop with Thread.sleep(200) **Test Failed for Windows platform**
2. Old test i.e. with SecondaryLoop with Thread.sleep(500) **Test Passed**

3. Changed test i.e without SecondaryLoop with added robot.delay(200) **Test Failed for Windows platform**
4. Changed test i.e. without SecondaryLoop with added robot.delay(500) **Test Passed**

Since the old test with longer delay (500 ms) and changed test with delay of 500 ms passed without the fix of JDK-8078269, looks like the reason for failure could be less delay.

prsadhuk commented:
I am not sure why secondary loop was added in the testcase for JDK-8078269..
JDK-8078269 issue was `TabbedPane.tabAreaBackground` should have precedence over `setBackground` but it was not working after JDK-8007563..
The 8007563 tests for non-opaque case whereas 8078269 issue was for opaque.
 It seems if we put `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN); ` then the tabbedpane areas where there are no tabs are not GREEN but RED...
Seems like we need a new testcase for JDK-8078269 and do away with secondary loop in 8007563 testcase..

kumarabhi006 commented:
> I am not sure why secondary loop was added in the testcase for JDK-8078269.. 

That is unclear to me as well. 

> JDK-8078269 issue was `TabbedPane.tabAreaBackground` should have precedence over `setBackground` but it was not working after JDK-8007563.. The 8007563 tests for non-opaque case whereas 8078269 issue was for opaque. It seems if we put `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN); ` then the tabbedpane areas where there are no tabs are not GREEN but RED... Seems like we need a new testcase for JDK-8078269 and do away with secondary loop in 8007563 testcase..

This issue still exists with the existing implementation in MetalTabbedPaneUI. If we add `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN); ` to the existing test then the expected output is that `the tabbedpane areas where there are no tabs should be GREEN`  but it is `RED` and the reason is the **background color instance check of UIResource** at [L905](https://github.com/kumarabhi006/jdk/blob/73f7a5f15dbba54a98f3916ff1190520ac07874d/src/java.desktop/share/classes/javax/swing/plaf/metal/MetalTabbedPaneUI.java#L905). Since `background is an instance of Color object` the condition evaluates to **false** and the graphic's object color is set to the background of component which is RED.

I think removal of SecondaryLoop should be ok for this PR. And for  JDK-8078269 as you suggested we may need a new test case which checks for the `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN);` as the property is not tested here.

prsadhuk commented:
> > I am not sure why secondary loop was added in the testcase for JDK-8078269..
> 
> That is unclear to me as well.
> 
> > JDK-8078269 issue was `TabbedPane.tabAreaBackground` should have precedence over `setBackground` but it was not working after JDK-8007563.. The 8007563 tests for non-opaque case whereas 8078269 issue was for opaque. It seems if we put `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN); ` then the tabbedpane areas where there are no tabs are not GREEN but RED... Seems like we need a new testcase for JDK-8078269 and do away with secondary loop in 8007563 testcase..
> 
> This issue still exists with the existing implementation in MetalTabbedPaneUI. If we add `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN); ` to the existing test then the expected output is that `the tabbedpane areas where there are no tabs should be GREEN` but it is `RED` and the reason is the **background color instance check of UIResource** at [L905](https://github.com/kumarabhi006/jdk/blob/73f7a5f15dbba54a98f3916ff1190520ac07874d/src/java.desktop/share/classes/javax/swing/plaf/metal/MetalTabbedPaneUI.java#L905). Since `background is an instance of Color object` the condition evaluates to **false** and the graphic's object color is set to the background of component which is RED.
> 
> I think removal of SecondaryLoop should be ok for this PR. And for JDK-8078269 as you suggested we may need a new test case which checks for the `UIManager.put("TabbedPane.tabAreaBackground", Color.GREEN);` as the property is not tested here.

As per
https://mail.openjdk.org/pipermail/swing-dev/2015-May/004521.html
secondary loop was added for
> Also inappropriate design chosen for the 8007563 reg test code affected test results stability. It also has been fixed.

I guess because pre-2018 our tests was run in samevm mode so there might be some stability issue and since it has been changed to othervm mode by default since '18, I guess this loop is not needed.
Also, as per the discussion for the webrev, it seems the "background" color, if set, was given precedence over property or in other words if `background color instance check of UIResource` means if background color is not set by user and is set by L&F which in case of MetalL&F is `"TabbedPane.background", controlShadow,` which is UIResource and in this case if `TabbedPane.tabAreaBackground` is set, TabbedPane.tabAreaBackground will take precedence

but if background Color set by user as is in this case in the testcase, it will not be instance of UIResource but Color so background color will take precedence

DamonGuy reviewed (changes requested):
I tested this myself and it does seem to be stable and pass consistently. Looks like the extra delay works.

Following the logic regarding the precedence of the different sources of background color, it does seem like Prasanta is correct when glancing over the different sources of colors.

Worth noting, I think it'd be good to rename the test since we're removing it from the problemlist and moving the test to a new location anyway. Maybe rename it to `JTabbedPaneBackgroundColorTest` or something similar to be more descriptive.

kumarabhi006 commented:
> Worth noting, I think it'd be good to rename the test since we're removing it from the problemlist and moving the test to a new location anyway. Maybe rename it to `JTabbedPaneBackgroundColorTest` or something similar to be more descriptive.

I didn't pay attention that the file is recognized as a new file when it has moved to new location. This may be due to the recent changes in test where it has been almost changed except few lines and is recognized as new file.
Renamed the test now.

prsadhuk commented:
Seems like closed test 4690946 where the UI property "TabbedPane.tabAreaBackground" was added in webrev.0 is moved to open and renamed as javax/swing/JTabbedPane/TestJTabbedPaneOpaqueColor.java..Any idea why the test is only for windows and linux? Can it be extended for macos too? Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?

kumarabhi006 commented:
> Seems like closed test 4690946 where the UI property "TabbedPane.tabAreaBackground" was added in webrev.0 is moved to open and renamed as javax/swing/JTabbedPane/TestJTabbedPaneOpaqueColor.java..Any idea why the test is only for windows and linux? Can it be extended for macos too? Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?

closed test 4690946 was moved to open for the issue [JDK-8226990](https://bugs.openjdk.org/browse/JDK-8226990) and was tested against the windows and linux platform only because the issue was reported for Nimbus and GTK L&F. [See the discussion here](https://github.com/openjdk/jdk/pull/17720#discussion_r1584537667).
For Aqua separate bug was filed [JDK-8331619](https://bugs.openjdk.org/browse/JDK-8331619) and [PR raised](https://github.com/openjdk/jdk/pull/19170). But this bug was closed as were unable to verify the native behavior for TabbedPane.

> Can it be extended for macos too?

Do you mean TestJTabbedPaneOpaqueColor Test ? I guess it can be extended for Mac but we need to restrict the testing only for Metal and Nimbus L&F on Mac machine.

kumarabhi006 commented:
> Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?

Yes, issue still exists for mac machine.

prsadhuk commented:
> > Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?
> 
> Yes, issue still exists for mac machine.

Issue is for AquaL&F which as per your verification in [JDK-8331619](https://bugs.openjdk.org/browse/JDK-8331619) and https://github.com/openjdk/jdk/pull/19170 is not an issue it seems..

kumarabhi006 commented:
> > > Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?
> > 
> > 
> > Yes, issue still exists for mac machine.
> 
> Issue is for AquaL&F which as per your verification in [JDK-8331619](https://bugs.openjdk.org/browse/JDK-8331619) and #19170 is not an issue it seems..

Yes, for Aqua L&F we were not able to verify the native behavior.
I think bug4690946.java test needs to be removed from closed problemlist.

prsadhuk commented:
> > > > Is [JDK-7148101](https://bugs.openjdk.org/browse/JDK-7148101) still valid?
> > > 
> > > 
> > > Yes, issue still exists for mac machine.
> > 
> > 
> > Issue is for AquaL&F which as per your verification in [JDK-8331619](https://bugs.openjdk.org/browse/JDK-8331619) and #19170 is not an issue it seems..
> 
> Yes, for Aqua L&F we were not able to verify the native behavior. I think bug4690946.java test needs to be removed from closed problemlist.

Yes already raised the PR for removal..

honkar-jdk commented on test/jdk/javax/swing/JTabbedPane/TestJTabbedPaneBackgroundColor.java:
```diff
@@ -0,0 +1,138 @@
+ */
+
+public class TestJTabbedPaneBackgroundColor {
+    private static final ArrayList<String> LIST = new ArrayList<>();
```
I think it can be changed to a non-final static var

```suggestion
    private static ArrayList<String> lafList = new ArrayList<>();
```

honkar-jdk commented on test/jdk/javax/swing/JTabbedPane/TestJTabbedPaneBackgroundColor.java:
```diff
@@ -0,0 +1,138 @@
+
+    public static void main(String[] args) throws Exception {
+        robot = new Robot();
+        robot.setAutoDelay(200);
```
This can be removed as there no Robot mouse or key events.
 `robot.setAutoDelay(200);`

kumarabhi006 replied:
Updated.

honkar-jdk reviewed (approved):
LGTM other than the minor test changes mentioned below.

honkar-jdk reviewed (approved):
LGTM.
Adding a consolidated note listing the JBS ids & corresponding tests related to JTabbedPane background issue in this PR or JBS might be helpful for future reference (avoids scrolling through PR comments).

1. [JDK-8051591](https://bugs.openjdk.org/browse/JDK-8051591) - Test javax/swing/JTabbedPane/8007563/Test8007563.java => test/jdk/javax/swing/JTabbedPane/TestJTabbedPaneBackgroundColor.java
...
so on.

kumarabhi006 commented:
/integrate
